integration/server.ts

```diff
@@ -1,48 +1,79 @@
-} = require('../dist-integration-server/main');
+// The Express app is exported so that it can be used by serverless Functions.
+export function app() {
+  const server = express();
+  const distFolder = join(process.cwd(), 'dist-integration');
+  const indexHtml = existsSync(join(distFolder, 'index.original.html'))
+    ? 'index.original.html'
+    : 'index';
 
-const PORT = process.env.PORT || 4200;
-const DIST_FOLDER = join(__dirname, '../dist-integration');
+  // Our Universal express-engine (found @ https://github.com/angular/universal/tree/master/modules/express-engine)
+  server.engine(
+    'html',
+    ngExpressEngine({
+      bootstrap: AppServerModule
+    }) as any
+  );
 
-enableProdMode();
+  server.set('view engine', 'html');
+  server.set('views', distFolder);
 
-const app = express();
-// Read `index.html` only once and cache it
-const document = readFileSync(join(DIST_FOLDER, 'index.html')).toString();
+  // Example Express Rest API endpoints
+  // app.get('/api/**', (req, res) => { });
+  // Serve static files from /browser
+  server.get(
+    '*.*',
+    express.static(distFolder, {
+      maxAge: '1y'
+    })
+  );
+
+  // All regular routes use the Universal engine
+  server.get('*', (req, res) => {
+    res.render(indexHtml, {
+      req,
+      providers: [
+        { provide: APP_BASE_HREF, useValue: req.baseUrl },
+        { provide: REQUEST, useValue: req },
+        { provide: RESPONSE, useValue: res }
+      ]
+    });
+  });
 
-// `index: false` means to ignore plain `index.html` thus
-// the render responsibility is fully taken by Angular Universal
-app.use(express.static(DIST_FOLDER, { index: false }));
+  return server;
+}
 
-app.get('*', async (req, res) => {
-  const url = req.url;
-  // tslint:disable-next-line:no-console
-  console.time(`GET: ${url}`);
+function run() {
+  const port = process.env.PORT || 4200;
 
-  const html = await renderModuleFactory(AppServerModuleNgFactory, {
-    url,
-    document,
-    extraProviders: [provideModuleMap(LAZY_MODULE_MAP)]
+  // Start up the Node server
+  const server = app();
+  server.listen(port, () => {
+    console.log(`Node Express server listening on http://localhost:${port}`);
   });
+}
 
-  // tslint:disable-next-line:no-console
-  console.timeEnd(`GET: ${url}`);
-  res.send(html);
-});
+// Webpack will replace 'require' with '__webpack_require__'
+// '__non_webpack_require__' is a proxy to Node 'require'
+// The below code is to ensure that the server is run only when not requiring the bundle.
+// eslint-disable-next-line
+declare const __non_webpack_require__: NodeRequire;
+// eslint-disable-next-line
+const mainModule = __non_webpack_require__.main;
+const moduleFilename = (mainModule && mainModule.filename) || '';
+if (moduleFilename === __filename || moduleFilename.includes('iisnode')) {
+  run();
+}
 
-app.listen(PORT, () => {
-  console.log(`Express server is running and listening at http://localhost:${PORT}!`);
-});
+export * from './main.server';
```

Comment:
Same question here, I don't see it imported anywhere...?

Reply:
It needed for Angular Ivy and SSR + ViewEngine
Otherwise, it just doesn't start and doesn't work.

Official example:
https://github.com/angular/universal/blob/master/integration/hapi-engine-ve/server.ts#L69

Another examples:
https://github.com/Angular-RU/angular-universal-starter/blob/master/server.ts#L189
https://www.ganatan.com/tutorials/server-side-rendering-with-angular-universal